### ISSUE TITLE
Test framework quit unexpectedly

### ISSUE THREAD

**Issue by JoeBruner** [go-lang-plugin-org/go-lang-idea-plugin#1379]
Using Alpha#134

If I right click inside a test and do `Run TestFoo...` it will run the test as expect.  I'll see in the terminal window:

    /usr/local/go/bin/go test -v myproj/subdir --run='^TestFoo&'
    Test started at 4:20 PM ...
    PASS
    ok    myproj/subdir  0.006s

However on the left, I get the error message

    Test framework quit unexpectedly

instead of the expected

    All Tests Passed

**Comment by dlsniper** [comment]
@JoeBruner can you please provide some more context for this?
What did you set as Go libraries under ``` Settings | Languages and frameworks | Go ```?
Where is your project dir in relation to the entries in the libraries section?

Thank you.

**Comment by zolotov** [comment]
@JoeBruner please try this build https://www.dropbox.com/s/eplutjm8ow4wx16/Go.148.zip?dl=0. I fixed single test running mode just after Go#134 have been published.

**Comment by zolotov** [comment]
@JoeBruner please ping me if it doesn't work for you, I'll reopen the issue.

**Comment by JoeBruner** [comment]
@zolotov It seems to be working right, thought there is one minor annoyance.  If I right click run the tests individually they work as expected, but if I right click between functions and try `Run my_test.go` it will fail with `Test framework quite unexpectedly`.  It will fail with the following:

    /usr/local/go/bin/go test -v /home/dev/projects/go/src/A/B/my_test.go
    Testing started at 7:13 AM ...
    # command-line-arguments
    ../B/my_test.go:89: undefined: Foo
    ../B/my_test.go:125: undefined: Foo
    ../B/my_test.go:134: undefined: CONST_T
    ../B/my_test.go:134: undefined: CONST_V
    ../B/my_test.go:139: undefined: CONST_T
    ../B/my_test.go:147: undefined: CONST_T
    ../B/my_test.go:147: undefined: CONST_V
    
    ../B/my_test.go:152: undefined: CONST_T
    ../B/my_test.go:173: undefined: CONST_V
    ../B/my_test.go:182: undefined: CONST_T
    ../B/my_test.go:182: too many errors
    FAIL	command-line-arguments [build failed]

This is to be expected since you can't run `go test` directly on a file unless you have different packages. Doing the above will create a test configuration called `my_test.go` which is set to `Test kind: File`.  If I switch that to `Package` and change the `Package: ` to `A/B` (in my case) with a `Pattern` of `Foo` I can actually run all of the tests.  This of course requires that all of my tests contain the name `Foo` and only in that one file.  It's not really intuitve, but it's shortcoming of `go test` and not the plugin.

**Comment by zolotov** [comment]
Yes, it's known issue. I don't like current behaviour too, but...

> This of course requires that all of my tests contain the name Foo and only in that one file.

I don't know how to make `go test` use tests from single file only. If you'll think of the solution, please let me know and I'll fix the File run configuration.

**Comment by zolotov** [comment]
Although, I think can collect all test-function names and make unmodifiable filter like 'Test1|Test2|...'

**Comment by JoeBruner** [comment]
Hmmm, that's intriguing.  I like it. The test names would have to be unique across files otherwise go will complain.  I think that might just work.  Brilliant!

**Comment by dlsniper** [comment]
@zolotov you can check this [stackoverflow answer](http://stackoverflow.com/a/16936314/951414) about running tests from a single file. The approach to get all tests functions seems good but there's going to be a need to change the filter every time there's a function added or removed.

**Comment by zolotov** [comment]
@dlsniper thanks for the link.

> The approach to get all tests functions seems good but there's going to be a need to change the filter every time there's a function added or removed.

Not really, I won't store the function names and will calculate it before each running

**Comment by dlsniper** [comment]
:+1:

**Comment by zolotov** [comment]
And here we are https://www.dropbox.com/s/eh0zvnga1eqznbb/Go.150.zip?dl=0

Please try it

**Comment by JoeBruner** [comment]
Works great.  I created test, ran it, added new function, re-ran and it picked up the new test.  I even right clicked on a folder and did `Run all in foo` and it worked.  Great work!

**Comment by rubens21** [comment]
For whoever came here after gets crazy because that message!

I got the same message because forgot to call `m.Run()` inside `TestMain` function.
:p

**Comment by stippi2** [comment]
I am experiencing this exact issue from 2015. I am not 100% sure, but I think it is after updating GoLand to 2020.1 yesterday. I don't understand the comment by @rubens21, for me the problem is with tests that worked fine before.

**Comment by rubens21** [comment]
Hi @stippi2 
I was implementing the function `TestMain`, that controls the test setup https://golang.org/pkg/testing/#hdr-Main).
We must call `m.Run()` inside `TestMain`, that's how the tests are fired.

I had forgotten to call it, and the error output was that one.

**Comment by stippi2** [comment]
Hi @rubens21,
thanks for replying! I have tried a few things, but in the end, I downgraded to GoLand 2019.3 and that solved it, all else being equal.
To me it looked a bit like GoLand was not compiling and handing the current version of the code to go when testing, but an earlier version. Tests which existed before my changes were running and passing. When I deliberately changed test expectations, the tests would still pass. New tests were seemingly not actually run (although "passing"), and execution would not stop at breakpoints that I installed. Interestingly, the log contained entries to the effect of "cannot install break point at line without statement". Hope this helps anyone investigating.
